### ISSUE TITLE
Realtime.js getting invalid token on disconnect/reconnect of websocket after gotrue.js token refresh

### ISSUE THREAD

**Issue by GaryAustin1** [supabase/realtime-js#123]
# Bug report
EDIT:
SEE 2nd REPLY comment below.   This is a separate bug from the background throttle after more investigation.  Any disconnect of the websocket after a token refresh by gotrue, and then a reconnect gets a token error.  The reconnect uses the original token provided to realtime rather than the more recent one provided by gotrue using realtime.setAuth().  It appears setAuth() only works in the context of the current websocket.  It needs to also change the token in the parameters passed to realtime at the start (which appear to be used to start a new websocket after disconnect).

--------------------

Related to my other report of background tabs causing realtime.js to lose connection regularly. 
It appears that background throttling on at least Chrome and Edge cause an issue between gotrue doing a token refresh request and a closely followed realtime reconnect or join to have a bad token.

I believe the problem is related to:

**Intensive throttling**

> OK, here's the new bit in Chrome 88. Intensive throttling happens to timers that are scheduled when none of the minimal throttling or throttling conditions apply, and all of the following conditions are true:
>     The page has been hidden for more than 5 minutes.
>     The chain count is 5 or greater.
>     The page has been silent for at least 30 seconds.
>     WebRTC is not in use.
> In this case, the browser will check timers in this group once per minute. Similar to before, this means timers will batch together in these minute-by-minute checks. 

  https://developer.chrome.com/blog/timer-throttling-in-chrome-88/

Realtime is rejoining constantly because of throttling and then when gotrue's refresh timing is slightly messed up realtime also misses the new token.

## To Reproduce
Edit: add correct code

var supabase = supabase.createClient(SUPABASE_URL, SUPABASE_KEY)
console.log(supabase)
async function app () {
    let status = await supabase.auth.signIn({email: 'gabbott@xyz.com', password: 'TestPass'})
    const mySubscription = supabase
        .from('test')
        .on('*', payload => {
            console.log('Change received!', payload)
        })
        .subscribe((status, e) => {
            console.log('status', status, e)
        })
}
app()

Start up subscription to a table. Start Dev Console turn on networking and timestamps. Shrink browser or hide tab under others. Wait several hours (need 1 or sometimes more token refresh cycles.

Websocket will have disconnected and connect continuously and eventually after a token refresh network request the next websocket reconnect gets token error and seems to continue to fail in some sort of loop.

## Expected behavior

realtime/gotrue should work in background tabs.

## Screenshots

network token request response access token:

{"access_token":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJhdXRoZW50aWNhdGVkIiwiZXhwIjoxNjM5MTc3NjkzLCJzdWIiOiI4NjU2N2IwYS1jMzA2LTRjY2UtOGZkYi00ZGE3ZWYwNzk3NGYiLCJlbWFpbCI6ImdhYmJvdHRAZ3JhcGVwYWQuY29tIiwicGhvbmUiOiIiLCJhcHBfbWV0YWRhdGEiOnsicHJvdmlkZXIiOiJlbWFpbCIsInByb3ZpZGVycyI6WyJlbWFpbCJdfSwidXNlcl9tZXRhZGF0YSI6e30sInJvbGUiOiJhdXRoZW50aWNhdGVkIn0.H9IHX9-vG8TtJ6IxjUkux7BNlkmQj6sMPbNdQKgD4lY"

Next websocket reconnect with a join message:

  {"topic":"realtime:public:test","event":"phx_join", "payload":  "payload":{"user_token":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJhdXRoZW50aWNhdGVkIiwiZXhwIjoxNjM5MTc0MTUzLCJzdWIiOiI4NjU2N2IwYS1jMzA2LTRjY2UtOGZkYi00ZGE3ZWYwNzk3NGYiLCJlbWFpbCI6ImdhYmJvdHRAZ3JhcGVwYWQuY29tIiwicGhvbmUiOiIiLCJhcHBfbWV0YWRhdGEiOnsicHJvdmlkZXIiOiJlbWFpbCIsInByb3ZpZGVycyI6WyJlbWFpbCJdfSwidXNlcl9tZXRhZGF0YSI6e30sInJvbGUiOiJhdXRoZW50aWNhdGVkIn0.fr6oHQsB6zzWBH0qmCWoLtUx35LwiFFI-q2iebJcW-E"},"ref":"207"}

same websocket reply
  {"event":"phx_reply","payload":{"response":{"reason":"user token is invalid"},"status":"error"},"ref":"207","topic":"realtime:public:test"}

NOTE THE TOKENS ARE NOT THE SAME. The websocket join request is using the previous token before the refresh call.

## System information
Windows 10, (Chrome, Edge latest with tab throttle disconnect) (any browser with offline/online after refresh token)
Latest supabase server and 1.28.5 supabase.js

## Additional context
I no longer think this is a multiuser problem, but a timing problem with throttling realtime to not catch the new token from the immediate previous token refresh call.  This is likely caused by the timer delays in Chrome based browsers.  I've not test Safari at all, Firefox seems fine.

**Comment by GaryAustin1** [comment]
More info on error sequence: 3 letter code is last 3 letters of token, times are in websocket messages except the gotrue net request.

Many websocket connections before failure....

websocket start:
31:03 -- join fmU
31:31 -- fmU request
-------------------------------31:51 (time from network packet) token request from gotrue, returns Djk token
31:52 -- DjK request           (note odd time not normal 30 sec heart beat)
32:01 -- DjK request
32:31 -- DjK req
33:01 -- DjK req
33:58 -- DjK req          (note now 57 seconds)
34:58 -- DjK req          (note now 60 seconds)
35:58 -- DjK req         (note still 60 seconds, right on edge of losing heart beat....)
websocket drops

new websocket starts
37:03 -- join with fmU       (back to old token)
37:03 **return from server token error**
continues on now with join requests on fmU token errors with random time intervals and thru multiple websocket connections...

**Comment by GaryAustin1** [comment]
Further testing reveals Firefox also fails and the failure only requires a running realtime websocket, then a refresh token by gotrue.  After that a disconnection of the websocket (background throttling, internet interruption, browser offline/online toggle) causes the reconnect to pass the original web token from realtime startup versus the latest token from gotrue refresh token.

I believe the realtime.setAuth(token) called by gotrue only updates the token in the context of the running websocket.  When the websocket terminates and reconnects for any reason, realtime is uses the original token provided at initialization.

**Comment by w3b6x9** [comment]
@GaryAustin1

> The reconnect uses the original token provided to realtime rather than the more recent one provided by gotrue using realtime.setAuth().

I think you're right! I'm currently putting in a fix.

**Comment by w3b6x9** [comment]
@GaryAustin1 if you're able to can you spin up these [changes](https://github.com/supabase/realtime-js/pull/126) locally and see if it fixes this issue you reported?

**Comment by GaryAustin1** [comment]
I'll give it a try.  I seemed to only half succeed last time, but I need to get this github stuff down.  I believe I have  pulled, and built the branch and started code running in my main project, now have to wait for token refresh.  If I don't respond tonight, will tomorrow.  

By the way my Supabase instance we started all this with seems to have partly switched back to the new RLS stuff (which is good) but I can't get any realtime events down the websocket on changes (RLS table or not).  A new spin up test case works fine.  About to not look at stuff over holiday, so no big deal at moment.  Flipped all the switches, etc.  I still see people out there having issues getting events back in discord so maybe something still strange on specific backends...

**Comment by w3b6x9** [comment]
@GaryAustin1 

> I'll give it a try.

Thanks!

> By the way my Supabase instance we started all this with seems to have partly switched back to the new RLS stuff (which is good) but I can't get any realtime events down the websocket on changes (RLS table or not).

It should be working now! There was an issue with your Realtime replication slot. The previous Realtime we rolled you back to used a permanent slot while the new Realtime RLS uses a temporary slot. Because we had manually rolled you back the slot didn't become active (permanent slot persisted) after upgrading you to the latest Realtime RLS. This shouldn't happen again for future upgrades because slot is and will continue to be temporary.

> About to not look at stuff over holiday

Thanks for all your help so far. Enjoy the holidays!
